File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque

def solution(s):
    answer = 0
    n = len(s)
    check = deque(s)
    
    for i in range(n):
        stack = []
        for j in range(n):
            c = check[j]
            if c == "(" or c == "{" or c == "[":
                stack.append(c)
            else:
                if not stack:
                    break
                if c == ")" and stack[-1] == "(":
                    stack.pop()
                elif c == "]" and stack[-1] == "[":
                    stack.pop()
                elif c == "}" and stack[-1] == "{":
                    stack.pop()
                else:
                    break
        else:
            if not stack:
                answer += 1
        check.rotate(-1)
    
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    let answer = 0;
    const n = s.length;

    // 괄호 짝을 맞추는 함수
    function isValid(str) {
        const stack = [];
        for (let i = 0; i < str.length; i++) {
            const ch = str[i];
            if (ch === '(' || ch === '{' || ch === '[') {
                stack.push(ch);
            } else {
                if (stack.length === 0) return false;
                const last = stack.pop();
                if (
                    (ch === ')' && last !== '(') ||
                    (ch === '}' && last !== '{') ||
                    (ch === ']' && last !== '[')
                ) {
                    return false;
                }
            }
        }
        return stack.length === 0;
    }

    // 문자열을 한 칸씩 회전
    for (let i = 0; i < n; i++) {
        const rotated = s.slice(i) + s.slice(0, i);
        if (isValid(rotated)) answer++;
    }

    return answer;
}
